crates/swc_css_minifier/src/compressor/calc_sum.rs

```diff
@@ -0,0 +1,1083 @@
+// a Vec<CalcProductOrOperator>
+#[derive(Debug, Clone)]
+struct IndexedData<T> {
+    pos: usize,
+    data: T,
+}
+
+#[derive(Debug, Clone)]
+struct IndexedOperatorAndOperand<T> {
+    operator: Option<IndexedData<CalcOperator>>,
+    operand: IndexedData<T>,
+}
+
+type SumOperation = fn(v1: f64, v2: f64, ratio: Option<f64>) -> f64;
+
+const SUM_OPERATION_ADD: SumOperation = |v1, v2, ratio| match ratio {
+    None => v1 + v2,
+    Some(r) => v2.mul_add(r, v1),
+};
+const SUM_OPERATION_SUBTRACT: SumOperation = |v1, v2, ratio| match ratio {
+    None => v1 - v2,
+    Some(r) => -(v2.mul_add(r, -v1)),
+};
+
+fn get_precision(n: f64) -> usize {
+    let n_as_str = n.to_string();
+    n_as_str.find('.').map_or(0, |sep| n_as_str.len() - sep)
+}
+
+#[derive(Default)]
+struct CalcSumContext {
+    number: Option<IndexedOperatorAndOperand<Number>>,
+    percentage: Option<IndexedOperatorAndOperand<Percentage>>,
+    absolute_length: Option<IndexedOperatorAndOperand<Length>>,
+    other_lengths: HashMap<String, IndexedOperatorAndOperand<Length>>,
+    angle: Option<IndexedOperatorAndOperand<Angle>>,
+    duration: Option<IndexedOperatorAndOperand<Time>>,
+    frequency: Option<IndexedOperatorAndOperand<Frequency>>,
+    resolution: Option<IndexedOperatorAndOperand<Resolution>>,
+    flex: Option<IndexedOperatorAndOperand<Flex>>,
+    unknown_dimension: HashMap<String, IndexedOperatorAndOperand<UnknownDimension>>,
+    expressions: Vec<CalcProductOrOperator>,
+}
+
+impl CalcSumContext {
+    pub fn fold(&mut self, calc_sum: &mut CalcSum) {
+        self.nested_fold(None, calc_sum);
+        calc_sum.expressions = self.expressions.to_vec();
+        remove_unnecessary_nesting_from_calc_sum(calc_sum);
+    }
+
+    fn nested_fold(&mut self, surrounding_operator: Option<&CalcOperator>, calc_sum: &mut CalcSum) {
+        let mut operator: Option<CalcOperator> =
+            CalcSumContext::merge_operators(surrounding_operator, None);
+
+        let mut expr_it = calc_sum.expressions.iter_mut();
+        while let Some(calc_product_or_operator) = expr_it.next() {
+            match calc_product_or_operator {
+                CalcProductOrOperator::Product(calc_product) => {
+                    fold_calc_product(calc_product);
+                    self.reduce(operator.as_ref(), calc_product);
+                }
+                _ => {
+                    // We should have an operand
+                    return;
+                }
+            };
+
+            if let Some(CalcProductOrOperator::Operator(op)) = expr_it.next() {
+                operator = CalcSumContext::merge_operators(surrounding_operator, Some(op));
+            }
+        }
+    }
+
+    fn merge_operators(
+        surrounding_operator: Option<&CalcOperator>,
+        nested_operator: Option<&CalcOperator>,
+    ) -> Option<CalcOperator> {
+        match nested_operator {
+            None => surrounding_operator.map(|so| (*so).clone()),
```

Comment:
I think `surrounding_operator.cloned()` is better than `surrounding_operator.map(|so| (*so).clone())`

Reply:
You're right, done